ConcurrentModificationException during communication channel close leaves serial ports busy

## Environment

| Component  | Details                                                |
|------------|--------------------------------------------------------|
| QZ Tray    | 2.2.5 (Java)                                           |
| OS/Arch    | macOS 15.7.3 (Apple Silicon), Windows 11              |
| Client     | "qz-tray" npm version ^2.2.5                                       |
| Browsers   | Chrome 144.0.7559.133 (Mac OS), Edge (Win 11)         |


## Steps to Reproduce

1. Start QZ Tray.
2. Web app connects via WebSocket.
3. Open serial Port A with:
`dataBits: 8, stopBits: 1, baudRate: 2400, parity: 'NONE', flowControl: 'NONE'`
4. Open serial Port B with:
`dataBits: 8, stopBits: 1, baudRate: 9600, parity: 'NONE', flowControl: 'NONE'` 
⚠️ The issue occurs only when more than one port is open.
5. Refresh the web page.
6. Observe QZ Tray logs

## Expected behavior
WebSocket closes cleanly; serial channels are released; ports can be reopened after refresh.

## Actual behavior
After refresh, opening at least one port fails with:
`Error: Port name - /dev/cu.usbserial-BZEEb114J20; Method name - openPort(); Exception type - Port busy.`

⚠️  Requires exiting & restarting QZ Tray app to recover.

## Reproducibility

Always (9/10 attempts)

## Workarounds tried

Full QZ Tray restart (works until next refresh). 

## Logs

```
2026-02-11T19:33:13,832 [DEBUG] Message: {"call":"serial.sendData","params":{"port":"\/dev\/cu.usbserial-1410","data":{"type":"HEX","data":"x02x4Fx03x0Dx0A"}},"signature":"***","timestamp":1770834793805,"uid":"thrszp","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:13,835 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:13,835 [DEBUG] Sending data over [/dev/cu.usbserial-1410]
2026-02-11T19:33:13,988 [INFO] Connection closed: 1001 - null
2026-02-11T19:33:13,988 [INFO] Closing all communication channels for ***
2026-02-11T19:33:13,988 [INFO] Picked up property tray.notifications=false
2026-02-11T19:33:14,492 [INFO] Serial port [/dev/cu.usbserial-1410] closed successfully.
2026-02-11T19:33:14,492 [ERROR] Failed to close communication channel
java.util.ConcurrentModificationException: null
	at java.base/java.util.HashMap$HashIterator.nextNode(Unknown Source) ~[?:?]
	at java.base/java.util.HashMap$ValueIterator.next(Unknown Source) ~[?:?]
	at qz.ws.SocketConnection.disconnect(SocketConnection.java:137) ~[qz-tray.jar:?]
	at qz.ws.PrintSocketClient.onClose(PrintSocketClient.java:81) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.notifyOnClose(JettyWebSocketFrameHandler.java:322) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.onCloseFrame(JettyWebSocketFrameHandler.java:281) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.common.JettyWebSocketFrameHandler.onFrame(JettyWebSocketFrameHandler.java:232) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession$IncomingAdaptor.onFrame(WebSocketCoreSession.java:705) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.AbstractExtension.nextIncomingFrame(AbstractExtension.java:145) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension.nextIncomingFrame(PerMessageDeflateExtension.java:236) ~[qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.DemandingFlusher.emitFrame(DemandingFlusher.java:145) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension$IncomingFlusher.handle(PerMessageDeflateExtension.java:378) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.DemandingFlusher.process(DemandingFlusher.java:169) [qz-tray.jar:?]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:243) [qz-tray.jar:?]
	at org.eclipse.jetty.util.IteratingCallback.succeeded(IteratingCallback.java:371) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.DemandingFlusher.onFrame(DemandingFlusher.java:104) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.PerMessageDeflateExtension.onFrame(PerMessageDeflateExtension.java:93) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.ExtensionStack.onFrame(ExtensionStack.java:120) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.WebSocketCoreSession.onFrame(WebSocketCoreSession.java:481) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFrame(WebSocketConnection.java:271) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.fillAndParse(WebSocketConnection.java:464) [qz-tray.jar:?]
	at org.eclipse.jetty.websocket.core.internal.WebSocketConnection.onFillable(WebSocketConnection.java:349) [qz-tray.jar:?]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314) [qz-tray.jar:?]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100) [qz-tray.jar:?]
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:421) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:390) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:277) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:199) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:411) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:969) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1194) [qz-tray.jar:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1149) [qz-tray.jar:?]
	at java.base/java.lang.Thread.run(Unknown Source) [?:?]
2026-02-11T19:33:15,698 [INFO] Connection opened from /0:0:0:0:0:0:0:1:59504 on socket port 8182
2026-02-11T19:33:15,698 [INFO] Picked up property tray.notifications=false
2026-02-11T19:33:15,699 [DEBUG] Message: {"call":"getVersion","timestamp":1770834795698,"uid":"cqjboz","position":{"x":1020,"y":-527.5}}
2026-02-11T19:33:15,700 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,707 [DEBUG] Message: {"certificate":"...","timestamp":1770834795707,"uid":"5rb0jb","position":{"x":1020,"y":-527.5}}
2026-02-11T19:33:15,710 [DEBUG] Successfully chained certificate: CN=***, O=*** (***)
2026-02-11T19:33:15,711 [WARN] Failed to retrieve QZ CRL, skipping CRL check
2026-02-11T19:33:15,712 [DEBUG] Received new certificate from connection through 59504
2026-02-11T19:33:15,792 [DEBUG] Message: {"call":"printers.detail","signature":"***","timestamp":1770834795722,"uid":"3bfuek","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,793 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,793 [DEBUG] Message: {"call":"serial.findPorts","signature":"***","timestamp":1770834795722,"uid":"3r5fsf","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,794 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,795 [DEBUG] Message: {"call":"serial.findPorts","signature":"***","timestamp":1770834795722,"uid":"fuwu2u","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,806 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,833 [DEBUG] Message: {"call":"serial.openPort","params":{"port":"\/dev\/cu.usbserial-BZEEb114J20","options":{"rx":{},"dataBits":8,"stopBits":1,"baudRate":2400,"parity":"NONE","flowControl":"NONE"}},"signature":"***","timestamp":1770834795812,"uid":"b202sx","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,834 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,852 [DEBUG] Message: {"call":"serial.openPort","params":{"port":"\/dev\/cu.usbserial-1410","options":{"rx":{},"dataBits":8,"stopBits":1,"baudRate":9600,"parity":"NONE","flowControl":"NONE"}},"signature":"***","timestamp":1770834795832,"uid":"2am2ht","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,853 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,862 [DEBUG] Applying new response formatting
2026-02-11T19:33:15,863 [DEBUG] Received serial output: 
2026-02-11T19:33:15,886 [DEBUG] Message: {"call":"serial.sendData","params":{"port":"\/dev\/cu.usbserial-1410","data":{"type":"HEX","data":"x02x4Fx03x0Dx0A"}},"signature":"***","timestamp":1770834795866,"uid":"8n3q1u","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:15,887 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:15,888 [DEBUG] Sending data over [/dev/cu.usbserial-1410]
2026-02-11T19:33:16,024 [DEBUG] Found 3 printers
2026-02-11T19:33:16,024 [INFO] Gathering printer MediaTray information...
2026-02-11T19:33:16,054 [DEBUG] Message: {"call":"printers.startListening","params":{"printerNames":[null]},"signature":"***","timestamp":1770834796029,"uid":"yx889t","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:16,056 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:16,057 [DEBUG] Message: {"call":"printers.getStatus","timestamp":1770834796057,"uid":"wyeuju","position":{"x":1020,"y":-527.5}}
2026-02-11T19:33:16,058 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:16,220 [DEBUG] Message: {"call":"serial.sendData","params":{"port":"\/dev\/cu.usbserial-1410","data":{"type":"HEX","data":"x02x4Fx03x0Dx0A"}},"signature":"***","timestamp":1770834796193,"uid":"8q1y3i","position":{"x":1020,"y":-527.5},"signAlgorithm":"SHA512"}
2026-02-11T19:33:16,222 [INFO] Substitutions file missing, skipping: /Library/Application Support/qz/substitutions.json
2026-02-11T19:33:16,222 [DEBUG] Sending data over [/dev/cu.usbserial-1410]
```

1. The issue `java.util.ConcurrentModificationException: null` is now resolved ✅
I tested the build from:
https://github.com/qzind/tray/actions/runs/21922109033
on both macOS and Windows.

2. The issue `org.eclipse.jetty.util.StaticException: Closed` still persists ❌
I tested the following build on macOS:
https://github.com/qzind/tray/actions/runs/22076021827


My observations:

1. I no longer see the exception `java.io.IOException: org.eclipse.jetty.util.StaticException: Closed` in the QZ Tray logs.
2. There are some new (or previously unseen) errors and warnings in the QZ Tray logs:
```2026-02-17T18:18:43,447 [ERROR] Stream is closed, could not send message
2026-02-17T18:28:56,807 [ERROR] Stream is closed, could not send message
...
2026-02-17T18:29:02,829 [WARN] Serial port [/dev/cu.usbserial-BZEEb114J20] was not closed properly.
2026-02-17T18:29:02,863 [WARN] Serial port [null] is not open.
2026-02-17T18:29:02,863 [WARN] Stopping CUPS status server
```
3. I can reproduce the issue only when trying to open 2 serial ports.

--- a/src/qz/communication/DeviceIO.java
+++ b/src/qz/communication/DeviceIO.java
@@ -6,6 +6,7 @@ public interface DeviceIO extends DeviceListener {
 
     String getProductId();
 
+    DeviceOptions getDeviceOptions();
 
     void open() throws DeviceException;
 

--- a/src/qz/communication/H4J_HidIO.java
+++ b/src/qz/communication/H4J_HidIO.java
@@ -55,6 +55,11 @@ public class H4J_HidIO implements DeviceIO, DeviceListener {
         return UsbUtil.toHexString(device.getProductId());
     }
 
+    @Override
+    public DeviceOptions getDeviceOptions() {
+        return dOpts;
+    }
+
     public byte[] readData(int responseSize, Byte unused) throws DeviceException {
         byte[] response = new byte[responseSize];
 

--- a/src/qz/communication/PJHA_HidIO.java
+++ b/src/qz/communication/PJHA_HidIO.java
@@ -89,6 +89,11 @@ public class PJHA_HidIO implements DeviceIO {
         return UsbUtil.toHexString(deviceInfo.getProductId());
     }
 
+    @Override
+    public DeviceOptions getDeviceOptions() {
+        return dOpts;
+    }
+
     public byte[] readData(int responseSize, Byte unused) throws DeviceException {
         byte[] response = new byte[responseSize];
         if (dataBuffer.isEmpty()) {

--- a/src/qz/communication/SerialIO.java
+++ b/src/qz/communication/SerialIO.java
@@ -265,9 +265,6 @@ public class SerialIO implements DeviceListener {
      */
     @Override
     public void close() {
-        // Remove orphaned reference
-        websocket.removeSerialPort(portName);
-
         if (!isOpen()) {
             log.warn("Serial port [{}] is not open.", portName);
         }

--- a/src/qz/communication/SocketIO.java
+++ b/src/qz/communication/SocketIO.java
@@ -74,9 +74,6 @@ public class SocketIO implements DeviceListener {
 
     @Override
     public void close() {
-        // Remove orphaned reference
-        websocket.removeNetworkSocket(String.format("%s:%s", host, port));
-
         try {
             dataOut.close();
         } catch(IOException e) {

--- a/src/qz/communication/UsbIO.java
+++ b/src/qz/communication/UsbIO.java
@@ -78,6 +78,11 @@ public class UsbIO implements DeviceIO {
         return UsbUtil.toHexString(iface.getUsbInterfaceDescriptor().iInterface());
     }
 
+    @Override
+    public DeviceOptions getDeviceOptions() {
+        return dOpts;
+    }
+
     public byte[] readData(int responseSize, Byte endpoint) throws DeviceException {
         try {
             byte[] response = new byte[responseSize];

--- a/src/qz/utils/FileWatcher.java
+++ b/src/qz/utils/FileWatcher.java
@@ -11,11 +11,12 @@ import java.io.FileReader;
 import java.io.IOException;
 import java.io.RandomAccessFile;
 import java.nio.channels.ClosedChannelException;
-import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.*;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class FileWatcher {
 
@@ -24,8 +25,7 @@ public class FileWatcher {
     private static Thread watchThread;
     private static WatchService watchService;
 
-    private static HashSet<FileIO> fileIOs = new HashSet<>();
-
+    private static final Set<FileIO> fileIOs = ConcurrentHashMap.newKeySet();
 
     public synchronized static void startWatchThread() throws IOException {
         if (watchThread != null && watchThread.isAlive()) {
@@ -59,7 +59,7 @@ public class FileWatcher {
         watchThread.start();
     }
 
-    public static void registerWatch(FileIO fileIO) throws IOException {
+    public synchronized static void registerWatch(FileIO fileIO) throws IOException {
         fileIO.setWk(fileIO.getAbsolutePath().register(watchService,
                                                        StandardWatchEventKinds.ENTRY_MODIFY,
                                                        StandardWatchEventKinds.ENTRY_CREATE,
@@ -73,7 +73,7 @@ public class FileWatcher {
     }
 
 
-    private synchronized static void fileChanged(Path path, String fileName, String type) throws ClosedChannelException {
+    private static void fileChanged(Path path, String fileName, String type) throws ClosedChannelException {
         Path filePath = path.resolve(fileName);
         for(FileIO fio : fileIOs) {
             if (!fio.isMatch(fileName)) continue;
@@ -109,7 +109,7 @@ public class FileWatcher {
             if (listener.isReversed()) { raf.seek(raf.length() - bytes.length); }
             raf.readFully(bytes);
 
-            return new String(bytes, Charset.forName("UTF-8"));
+            return new String(bytes, StandardCharsets.UTF_8);
         }
     }
 
@@ -118,7 +118,7 @@ public class FileWatcher {
 
         String buffer;
         if (listener.isReversed()) {
-            try(ReversedLinesFileReader reader = new ReversedLinesFileReader(path.toFile())) {
+            try(ReversedLinesFileReader reader = ReversedLinesFileReader.builder().setPath(path).get()) {
                 int count = 0;
                 while((buffer = reader.readLine()) != null && count++ != listener.getLines()) {
                     // Warning, this will strip "\r" from the data

--- a/src/qz/utils/SerialUtilities.java
+++ b/src/qz/utils/SerialUtilities.java
@@ -263,6 +263,11 @@ public class SerialUtilities {
                         StreamEvent event = new StreamEvent(StreamEvent.Stream.SERIAL, StreamEvent.Type.RECEIVE)
                                 .withData("portName", portName).withData("output", output);
                         PrintSocketClient.sendStream(session, event, serial);
+
+                        if (!serial.isOpen()) {
+                            //error happened sending stream data, cleanup
+                            connection.removeSerialPort(portName);
+                        }
                     }
                 });
 

--- a/src/qz/utils/SocketUtilities.java
+++ b/src/qz/utils/SocketUtilities.java
@@ -64,6 +64,11 @@ public class SocketUtilities {
                         PrintSocketClient.sendStream(session, eventErr, socket);
                     }
 
+                    if (!socket.isOpen()) {
+                        // ensure cleanup if device was closed
+                        connection.removeNetworkSocket(String.format("%s:%s", host, port));
+                    }
+
                     try { Thread.sleep(100); } catch(Exception ignore) {}
                 }).start();
 

--- a/src/qz/utils/UsbUtilities.java
+++ b/src/qz/utils/UsbUtilities.java
@@ -223,6 +223,11 @@ public class UsbUtilities {
                                 } /** No-op prevents re-throw **/);
                             }
                         }
+
+                        if (!usb.isOpen()) {
+                            // ensure cleanup if device was closed
+                            connection.removeDevice(dOpts);
+                        }
                     }
                 }.start();
 

--- a/src/qz/ws/ClosedSocketException.java
+++ b/src/qz/ws/ClosedSocketException.java
@@ -1,0 +1,61 @@
+package qz.ws;
+
+import org.eclipse.jetty.util.StaticException;
+
+import java.io.IOException;
+import java.nio.channels.ClosedChannelException;
+import java.util.Locale;
+
+/**
+ * <code>ClosedChannelException</code> wrapper class for handling edge-cases where Jetty nests or hides
+ * the underlying <code>Exception</code> cause.  Inspired by #1300 and #1414
+ */
+public class ClosedSocketException extends ClosedChannelException {
+    public ClosedSocketException(String message) {
+        super();
+        this.initCause(new IOException(message));
+    }
+
+    public ClosedSocketException(Throwable t) {
+        super();
+        this.initCause(t);
+    }
+
+    public static void filter(Throwable t) throws ClosedChannelException {
+        Throwable throwable = t;
+        while(true) {
+            if(throwable instanceof ClosedChannelException) {
+                throw (ClosedChannelException)throwable;
+            }
+            if(throwable != null) {
+                throwable = throwable.getCause();
+            } else {
+                break;
+            }
+        }
+
+        if (messageMatchesClosed(t)) {
+            throw new ClosedSocketException(t);
+        }
+    }
+
+    /**
+     * Check if a <code>Throwable</code> or its cause matches a <code>StaticException</code>
+     * with "closed" in the message string.
+     */
+    private static boolean messageMatchesClosed(Throwable t) {
+        Throwable throwable = t;
+        while(true) {
+            if(throwable == null) {
+                return false;
+            } else if(!(throwable instanceof StaticException)) {
+                // go deeper
+                throwable = throwable.getCause();
+            } else {
+                break;
+            }
+        }
+        StaticException staticException = (StaticException)throwable;
+        return staticException.getMessage() != null && staticException.getMessage().toLowerCase(Locale.ENGLISH).contains("closed");
+    }
+}

--- a/src/qz/ws/ClosedSocketException.java
+++ b/src/qz/ws/ClosedSocketException.java
@@ -24,6 +24,7 @@ public class ClosedSocketException extends ClosedChannelException {
     public static void filter(Throwable t) throws ClosedChannelException {
         Throwable throwable = t;
         while(true) {
+            // Look for nested ClosedChannelException
             if(throwable instanceof ClosedChannelException) {
                 throw (ClosedChannelException)throwable;
             }
@@ -34,6 +35,7 @@ public class ClosedSocketException extends ClosedChannelException {
             }
         }
 
+        // Look for edge-cases, such as StaticException("Closed");
         if (messageMatchesClosed(t)) {
             throw new ClosedSocketException(t);
         }

--- a/src/qz/ws/ClosedSocketException.java
+++ b/src/qz/ws/ClosedSocketException.java
@@ -21,6 +21,10 @@ public class ClosedSocketException extends ClosedChannelException {
         this.initCause(t);
     }
 
+    /**
+     * Rethrows anything that we believe to be related to a <code>ClosedChannelException</code>
+     * Does nothing otherwise.
+     */
     public static void filter(Throwable t) throws ClosedChannelException {
         Throwable throwable = t;
         while(true) {

--- a/src/qz/ws/PrintSocketClient.java
+++ b/src/qz/ws/PrintSocketClient.java
@@ -530,6 +530,7 @@ public class PrintSocketClient {
                 DeviceIO usb = connection.getDevice(dOpts);
                 if (usb != null) {
                     usb.close();
+                    connection.removeDevice(usb.getDeviceOptions());
 
                     sendResult(session, UID, null);
                 } else {

--- a/src/qz/ws/PrintSocketClient.java
+++ b/src/qz/ws/PrintSocketClient.java
@@ -828,14 +828,14 @@ public class PrintSocketClient {
      */
     private static synchronized void send(Session session, JSONObject reply) throws WebSocketException, ClosedChannelException {
         try {
-            session.getRemote().sendString(reply.toString());
+            if(session.isOpen()) {
+                session.getRemote().sendString(reply.toString());
+            } else {
+                throw new ClosedSocketException("Channel was closed before message could be sent");
+            }
         }
         catch(IOException e) {
-            if(e instanceof ClosedChannelException) {
-                throw (ClosedChannelException)e;
-            } else if(e.getCause() instanceof ClosedChannelException) {
-                throw (ClosedChannelException)e.getCause();
-            }
+            ClosedSocketException.filter(e); // Explicitly detect and re-throw ClosedChannelExceptions
             log.error("Could not send message", e);
         }
     }

--- a/src/qz/ws/SocketConnection.java
+++ b/src/qz/ws/SocketConnection.java
@@ -10,7 +10,7 @@ import qz.utils.FileWatcher;
 
 import java.io.IOException;
 import java.nio.file.Path;
-import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class SocketConnection {
 
@@ -19,18 +19,18 @@ public class SocketConnection {
 
     private Certificate certificate;
 
-    private DeviceListener deviceListener;
+    private volatile DeviceListener deviceListener;
 
     // serial port -> open SerialIO
-    private final HashMap<String,SerialIO> openSerialPorts = new HashMap<>();
+    private final ConcurrentHashMap<String,SerialIO> openSerialPorts = new ConcurrentHashMap<>();
     // socket 'host:port' -> open ProtocolIO
-    private final HashMap<String,SocketIO> openNetworkSockets = new HashMap<>();
+    private final ConcurrentHashMap<String,SocketIO> openNetworkSockets = new ConcurrentHashMap<>();
 
     // absolute path -> open file listener
-    private final HashMap<Path,FileIO> openFiles = new HashMap<>();
+    private final ConcurrentHashMap<Path,FileIO> openFiles = new ConcurrentHashMap<>();
 
     // DeviceOptions -> open DeviceIO
-    private final HashMap<DeviceOptions,DeviceIO> openDevices = new HashMap<>();
+    private final ConcurrentHashMap<DeviceOptions,DeviceIO> openDevices = new ConcurrentHashMap<>();
 
 
     public SocketConnection(Certificate cert) {
@@ -46,28 +46,28 @@ public class SocketConnection {
     }
 
 
-    public void addSerialPort(String port, SerialIO io) {
+    public synchronized void addSerialPort(String port, SerialIO io) {
         openSerialPorts.put(port, io);
     }
 
-    public SerialIO getSerialPort(String port) {
+    public synchronized SerialIO getSerialPort(String port) {
         return openSerialPorts.get(port);
     }
 
-    public void removeSerialPort(String port) {
+    public synchronized void removeSerialPort(String port) {
         openSerialPorts.remove(port);
     }
 
 
-    public void addNetworkSocket(String location, SocketIO io) {
+    public synchronized void addNetworkSocket(String location, SocketIO io) {
         openNetworkSockets.put(location, io);
     }
 
-    public SocketIO getNetworkSocket(String location) {
+    public synchronized SocketIO getNetworkSocket(String location) {
         return openNetworkSockets.get(location);
     }
 
-    public void removeNetworkSocket(String location) {
+    public synchronized void removeNetworkSocket(String location) {
         openNetworkSockets.remove(location);
     }
 
@@ -87,37 +87,38 @@ public class SocketConnection {
         deviceListener = null;
     }
 
-    public void addFileListener(Path absolute, FileIO listener) {
+    public synchronized void addFileListener(Path absolute, FileIO listener) {
         openFiles.put(absolute, listener);
     }
 
-    public FileIO getFileListener(Path absolute) {
+    public synchronized FileIO getFileListener(Path absolute) {
         return openFiles.get(absolute);
     }
 
-    public void removeFileListener(Path absolute) {
+    public synchronized void removeFileListener(Path absolute) {
         openFiles.remove(absolute);
     }
 
-    public void removeAllFileListeners() {
-        for(Path path : openFiles.keySet()) {
-            openFiles.get(path).close();
-            FileWatcher.deregisterWatch(openFiles.get(path));
+    public synchronized void removeAllFileListeners() {
+        for(FileIO io : openFiles.values()) {
+            if (io != null) {
+                io.close();
+                FileWatcher.deregisterWatch(io);
+            }
         }
-
         openFiles.clear();
     }
 
 
-    public void addDevice(DeviceOptions dOpts, DeviceIO io) {
+    public synchronized void addDevice(DeviceOptions dOpts, DeviceIO io) {
         openDevices.put(dOpts, io);
     }
 
-    public DeviceIO getDevice(DeviceOptions dOpts) {
+    public synchronized DeviceIO getDevice(DeviceOptions dOpts) {
         return openDevices.get(dOpts);
     }
 
-    public void removeDevice(DeviceOptions dOpts) {
+    public synchronized void removeDevice(DeviceOptions dOpts) {
         openDevices.remove(dOpts);
     }
 

--- a/src/qz/ws/SocketConnection.java
+++ b/src/qz/ws/SocketConnection.java
@@ -138,15 +138,18 @@ public class SocketConnection {
         for(SerialIO sio : openSerialPorts.values()) {
             sio.close();
         }
+        openSerialPorts.clear();
 
         for(SocketIO pio : openNetworkSockets.values()) {
             pio.close();
         }
+        openNetworkSockets.clear();
 
         for(DeviceIO dio : openDevices.values()) {
             dio.setStreaming(false);
             dio.close();
         }
+        openDevices.clear();
 
         removeAllFileListeners();
         stopDeviceListening();

--- a/src/qz/ws/SocketConnection.java
+++ b/src/qz/ws/SocketConnection.java
@@ -10,7 +10,7 @@ import qz.utils.FileWatcher;
 
 import java.io.IOException;
 import java.nio.file.Path;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.HashMap;
 
 public class SocketConnection {
 
@@ -22,15 +22,15 @@ public class SocketConnection {
     private volatile DeviceListener deviceListener;
 
     // serial port -> open SerialIO
-    private final ConcurrentHashMap<String,SerialIO> openSerialPorts = new ConcurrentHashMap<>();
+    private final HashMap<String,SerialIO> openSerialPorts = new HashMap<>();
     // socket 'host:port' -> open ProtocolIO
-    private final ConcurrentHashMap<String,SocketIO> openNetworkSockets = new ConcurrentHashMap<>();
+    private final HashMap<String,SocketIO> openNetworkSockets = new HashMap<>();
 
     // absolute path -> open file listener
-    private final ConcurrentHashMap<Path,FileIO> openFiles = new ConcurrentHashMap<>();
+    private final HashMap<Path,FileIO> openFiles = new HashMap<>();
 
     // DeviceOptions -> open DeviceIO
-    private final ConcurrentHashMap<DeviceOptions,DeviceIO> openDevices = new ConcurrentHashMap<>();
+    private final HashMap<DeviceOptions,DeviceIO> openDevices = new HashMap<>();
 
 
     public SocketConnection(Certificate cert) {
@@ -46,28 +46,28 @@ public class SocketConnection {
     }
 
 
-    public synchronized void addSerialPort(String port, SerialIO io) {
+    public void addSerialPort(String port, SerialIO io) {
         openSerialPorts.put(port, io);
     }
 
-    public synchronized SerialIO getSerialPort(String port) {
+    public SerialIO getSerialPort(String port) {
         return openSerialPorts.get(port);
     }
 
-    public synchronized void removeSerialPort(String port) {
+    public void removeSerialPort(String port) {
         openSerialPorts.remove(port);
     }
 
 
-    public synchronized void addNetworkSocket(String location, SocketIO io) {
+    public void addNetworkSocket(String location, SocketIO io) {
         openNetworkSockets.put(location, io);
     }
 
-    public synchronized SocketIO getNetworkSocket(String location) {
+    public SocketIO getNetworkSocket(String location) {
         return openNetworkSockets.get(location);
     }
 
-    public synchronized void removeNetworkSocket(String location) {
+    public void removeNetworkSocket(String location) {
         openNetworkSockets.remove(location);
     }
 
@@ -87,19 +87,19 @@ public class SocketConnection {
         deviceListener = null;
     }
 
-    public synchronized void addFileListener(Path absolute, FileIO listener) {
+    public void addFileListener(Path absolute, FileIO listener) {
         openFiles.put(absolute, listener);
     }
 
-    public synchronized FileIO getFileListener(Path absolute) {
+    public FileIO getFileListener(Path absolute) {
         return openFiles.get(absolute);
     }
 
-    public synchronized void removeFileListener(Path absolute) {
+    public void removeFileListener(Path absolute) {
         openFiles.remove(absolute);
     }
 
-    public synchronized void removeAllFileListeners() {
+    public void removeAllFileListeners() {
         for(FileIO io : openFiles.values()) {
             if (io != null) {
                 io.close();
@@ -110,15 +110,15 @@ public class SocketConnection {
     }
 
 
-    public synchronized void addDevice(DeviceOptions dOpts, DeviceIO io) {
+    public void addDevice(DeviceOptions dOpts, DeviceIO io) {
         openDevices.put(dOpts, io);
     }
 
-    public synchronized DeviceIO getDevice(DeviceOptions dOpts) {
+    public DeviceIO getDevice(DeviceOptions dOpts) {
         return openDevices.get(dOpts);
     }
 
-    public synchronized void removeDevice(DeviceOptions dOpts) {
+    public void removeDevice(DeviceOptions dOpts) {
         openDevices.remove(dOpts);
     }
 

